src/mpi/init/initthread.c

```diff
@@ -359,6 +359,45 @@ int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
             !hwloc_get_proc_cpubind(MPIR_Process.hwloc_topology, getpid(), MPIR_Process.bindset,
                                     HWLOC_CPUBIND_PROCESS);
     }
+
+    /* detect HBM nodes */
+    hwloc_obj_t obj;
+    MPIR_Process.hbmset_is_valid = 0;
+    int i, numa_nodes = hwloc_get_nbobjs_by_type(MPIR_Process.hwloc_topology, HWLOC_OBJ_NUMANODE);
+    MPIR_Process.hbm_nodes = hwloc_bitmap_alloc();
+    for (i = 0; i < numa_nodes; i++) {
+        obj = hwloc_get_obj_by_type(MPIR_Process.hwloc_topology, HWLOC_OBJ_NUMANODE, i);
+        if (obj->subtype && strcmp(obj->subtype, "MCDRAM") == 0) {
```

Comment:
Isn't this assuming that `HBM` is basically the same as `MCDRAM`?  It'd have been cleaner if you had a `hbm_type` that could default to `auto`, meaning whatever `hwloc` tells us is available on the node.  The user can force different placements when there are multiple other types of memories available.

In fact, it'd be even cleaner if we simply had `memory_type`.  `DRAM` is one memory type.  `MCDRAM` is another.  There might be more in the future.

Reply:
This is specific to MCDRAM since we didn't have other devices we could play with. That `if(obj->subtype && strcmp(obj->subtype, "MCDRAM")` can be extended with multiple `else if` to include other memories.

Reply:
I'm not asking you to test with more memories.  I'm just saying that your current code assumes that HBM == MCDRAM.  I think the detection code should be separated from the usage code.  For detection, we can detect all available forms of memory: DDR, MCDRAM, ...  For usage, we should default to DDR and if the user asks for something else, then we can see if such memory is available and use it.

Reply:
Ok I will think about how to integrate the changes to address you comment